docs/git.rst

```diff
@@ -0,0 +1,611 @@
+you will find plenty of literature.
+
+Reviewing your changes
+----------------------
+After hacking for some minutes/hours/days you are finished and about to commit your changes,
+great!
+
+*BUT*,
+please,
+do so with :command:`git add --patch`.
+
+The ``--patch`` (also ``-p``) switch allows you to select which hunks you want to add on a commit.
+
+This is not only great to split changes into different commits,
+but is also the time when you actually **review** your code before anyone else sees it.
+
+This is the time when you spot typos,
+pep8 errors,
+misaligned code,
+lack of docstrings in methods,
+that a permission is not defined on Generic Setup,
+that an upgrade should be needed...
+
+Remember that the first code review is the one you do on your own.
+Some inspiration/better phrased:
+http://ada.mbecker.cc/2012/11/22/be-your-own-code-review/
+
+And please,
+do remember the gold metric about reviewing code:
+http://www.osnews.com/story/19266/WTFs_m
+
+One commit does one thing
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Repeat with me:
+*One commit does one thing*.
+Period.
+
+When someone else needs to review your code,
+most probably she will give up or just skim over your code if there are too many (unrelated) changes.
+
+Reviewing commits with +20 files doing all sorts of changes on them
+(maybe even unrelated)
+is no fun and adds complexity and `cognitive load <http://en.wikipedia.org/wiki/Cognitive_load>`_.
+
+Something that should mostly be a verification of a checklist like:
+
+- the browser view is registered on ZCML?
+- is there an interface for that form?
+- the pt and py are there?
+- ...
+
+Turns instead into a list of questions:
+
+- why this interface is renamed here if has nothing to do with this adapter?
+- all this removal of deprecated code while adding new features just mixes the diff,
+  am I missing something?
+- *others*
+
+If you can not express what has been changed within 50 characters
+(suggested length of a commit message subject),
+or you say it like "it does XXX and YYY",
+you most probably need to split that commit into,
+at least,
+two or more commits.
+
+That doesn't mean that a +20 files or +100 lines of code changes are bad per se,
+you may be doing a simple refactoring across lots of files,
+that's fine and good actually.
+
+As long as a commit is just and only about a specific purpose,
+and not a mixed selection of the following:
+
+- refactoring code
+- moving things around
+- fixing some bugs while at it
+- adding some docs
+- a new cool feature
+- fixing typos on documentation
+- pep8 fixes
+
+Is completely fine to make refactorings.
+
+And this is actually to help both your present self and your +5 years from now that will have to refactor that code of yours,
+and maybe is struggling to understand what was going on there.
+
+Following this advice will:
+
+- keep things simple where there's no gain on adding complexity
+- make your changes easy to be reviewed
+- make later on lookups on those changes easy to follow
```

Comment:
Sorry for the late answer: I'm trying to say that if "one commit does one thing" if you ever have to do something similar (say create a viewlet) it will make your life easier as it will be a matter of copy/pasting that previous commit that you can look it up.

Does that make sense?